### ISSUE TITLE
Gradients are [None,...] in text GAN model

### ISSUE THREAD

**Issue by e8panse** [keras-team/keras#17444]
I'm trying to build text GAN using TF2. But gradients_of_generator always becoming [None].  Provided code below uses paddings in between generated_prediction and discriminator input. BUT even if I directly pass generator output to discriminator input (of course changing input_length of discriminator), gradients always [None].

How compute gradient in such text GAN correclty?

Please notice, I simplifed code to give more clear understanding of the problem.

code:

`generator = keras.Sequential(
    [
        layers.Embedding(total_words, 80, input_length=max_sequence_length-1),
        LSTM(100, return_sequences=True),
        LSTM(50),
        tf.keras.layers.Dropout(0.1),
        Dense(total_words/20),
        Dense(total_words, activation='softmax'),
    ],
    name="generator",
)

discriminator = keras.Sequential(
    [
        layers.Embedding(total_words, 80, input_length=max_sequence_length-1),
        layers.Bidirectional(tf.keras.layers.LSTM(64)),
        layers.Dropout(0.2),
        layers.Dense(64, activation='relu'),
        layers.Dense(1, activation='sigmoid'),
    ],
    name="discriminator",
)

    #function: train_step
    #text - tokenized text

    with tf.GradientTape() as gen_tape:
        
        predicted_next_word = generator(text, training=True)    
        pred_to_word = pad_sequences([[np.argmax(predicted_next_word, axis=-1)]], maxlen=max_sequence_length-1, padding='pre')
        fake_output = discriminator(pred_to_word, training=True)
        generator_loss = binary_cross_entropy(tf.ones(fake_output.shape), fake_output) #gives correct loss results
        
    gradients_of_generator = gen_tape.gradient(generator_loss, generator.trainable_variables)`

Gradient are [None,...]

**Comment by tilakrayal** [comment]
@e8panse,
I could see that the given code was incomplete. Kindly find the gist of it here and provide the complete code with the 
dependencies. Thank you!

**Comment by e8panse** [comment]
`
import time
import random
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.text import one_hot
import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.layers as layers
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Dropout
from tensorflow.keras.layers import LSTM
from tensorflow.keras.callbacks import ModelCheckpoint
from tensorflow.keras.utils import to_categorical
import numpy as np
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.callbacks import ModelCheckpoint

from mgan_1_data_loader import DKDataLoader

batch_size=64
learning_rate=0.01
#use fixed length to simplify test
fixed_sentence_length=30

data_loader = DKDataLoader()
input_sequences = data_loader.tokenization()
fake_sequences = data_loader.generate_fake_text()
xs,ys = data_loader.data_prepare(input_sequences)
total_words = data_loader.total_words #63,303
max_sequence_length = data_loader.max_sequence_length #568
#total samples 100,563

# Reserve 5,000 samples for validation.
x_val = xs[-5000:]
y_val = ys[-5000:]
x_train = xs[:-5000]
y_train = ys[:-5000]

# Prepare the training dataset.
real_train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
real_train_dataset = real_train_dataset.shuffle(buffer_size=1024).batch(batch_size)
real_train_dataset_iter = real_train_dataset.__iter__()

# Prepare the validation dataset.
val_dataset = tf.data.Dataset.from_tensor_slices((x_val, y_val))
val_dataset = val_dataset.batch(batch_size)

generator = keras.Sequential(
    [
        #keras.Input(shape=(max_sequence_length-1,)),
        layers.Embedding(total_words, 80, input_length=max_sequence_length-1),
        LSTM(100, return_sequences=True),
        LSTM(50),
        tf.keras.layers.Dropout(0.1),
        Dense(total_words/20),
        Dense(total_words, activation='softmax'),
    ],
    name="generator",
)

class GeneratorModel(tf.keras.Model):

    def __init__(self, total_words, max_sequence_length):
        super().__init__()
        #self.generator = generator
        self.l1 = tf.keras.layers.Embedding(total_words, 80, input_length=max_sequence_length-1)
        self.l11 = LSTM(100, return_sequences=True)
        self.l12 = LSTM(50)
        self.l2 = tf.keras.layers.Dropout(0.1)
        self.l3 = Dense(total_words/20)
        self.l4 = Dense(total_words, activation='softmax')

    def call(self, x):
        x = self.l1(x)
        x = self.l11(x)
        x = self.l12(x)
        x = self.l2(x)
        x = self.l3(x)
        x = self.l4(x)
        return x

generator = GeneratorModel(total_words, max_sequence_length)

discriminator = keras.Sequential(
    [
        #keras.Input(shape=(max_sequence_length-1,)),
        layers.Embedding(total_words, 80, input_length=max_sequence_length-1),
        layers.Bidirectional(tf.keras.layers.LSTM(64)),
        #layers.GlobalAveragePooling1D(),
        layers.Dropout(0.2),
        layers.Dense(64, activation='relu'),
        layers.Dense(1, activation='sigmoid'),
    ],
    name="discriminator",
)

def generate_text_from_word_string(generator_model, seed_text):
    for _ in range(fixed_sentence_length):
        token_list = data_loader.tokenizer.texts_to_sequences([seed_text])[0]
        token_list = pad_sequences([token_list], maxlen=max_sequence_length-1, padding='pre')
        pred_1 = generator_model(token_list, training=True)
        predicted = np.argmax(pred_1, axis=-1)
        output_word = ""
        for word, index in data_loader.tokenizer.word_index.items():
            if index == predicted:
                output_word = word
                #next_words = [""]
                break
        seed_text += " " + output_word

    return seed_text

def text_to_sequence(seed_text):
    token_list = data_loader.tokenizer.texts_to_sequences([seed_text])[0]
    token_list = pad_sequences([token_list], maxlen=max_sequence_length, padding='pre')
    return token_list


def generate_and_save_text(epoch, seed_text):
    with open('saved_generated_text/saved.txt', 'a') as the_file:
        the_file.write(seed_text+' ---epoch:'+str(epoch)+' \n')

optimizer_generator=keras.optimizers.Adam(learning_rate=0.0003)
optimizer_discriminator=keras.optimizers.Adam(learning_rate=0.0003)

loss_fn_generator=keras.losses.CategoricalCrossentropy()
loss_fn_discriminator=keras.losses.BinaryCrossentropy()

binary_cross_entropy = tf.keras.losses.BinaryCrossentropy(from_logits=True)

def discriminator_loss(real_output, fake_output):
    real_loss = binary_cross_entropy(tf.ones_like(real_output), real_output)
    fake_loss = binary_cross_entropy(tf.zeros_like(fake_output), fake_output)
    total_loss = real_loss + fake_loss
    return total_loss

def generator_loss(fake_output, fake_ys):
    return binary_cross_entropy(fake_ys, fake_output)

@tf.function
def train_step(data_batch):

    real_x,real_y = data_batch
    word_i = random.randint(0, total_words-1)
    tl_res = tf.convert_to_tensor(pad_sequences([[word_i]], maxlen=max_sequence_length-1, padding='pre'))

    with tf.GradientTape() as gen_tape:
        #gen_tape.watch(generator.trainable_variables)
        pred_1 = generator(tl_res, training=True)
        pred_to_word = tf.convert_to_tensor(pad_sequences([[25]], maxlen=max_sequence_length-1, padding='pre'))
        #even if I change discriminator input and pass generator output directly to dicsriminator it cannot count gradients
        fake_output = discriminator(pred_to_word, training=True)
        
        gen_loss = binary_cross_entropy(tf.ones(fake_output.shape), fake_output)
        

    gradients_of_generator = gen_tape.gradient(gen_loss, generator.trainable_variables)
    print(gradients_of_generator)
    
    optimizer_generator.apply_gradients(zip(gradients_of_generator, generator.trainable_variables))
    #optimizer_discriminator.apply_gradients(zip(gradients_of_discriminator, discriminator.trainable_variables))


def train(epochs):
    for epoch in range(epochs):
        start = time.time()

    for data_batch in real_train_dataset:
        train_step(data_batch)

    # Produce texts as you go
    seed = random.randint(0, total_words-1)
    word = str(data_loader.word_from_index(seed))
    g_text = generate_text_from_word_string(generator, word)
    generate_and_save_text(epoch + 1, g_text)

    # Save the model every 15 epochs
    if (epoch + 1) % 15 == 0:
        #checkpoint.save(file_prefix = checkpoint_prefix)
        pass

    print ('Time for epoch {} is {} sec'.format(epoch + 1, time.time()-start))

    # Generate after the final epoch
    seed = random.randint(0, total_words-1)
    word = str(data_loader.word_from_index(seed))
    g_text = generate_text_from_word_string(generator, word)
    generate_and_save_text(epochs, g_text)

train(2)
`

**Comment by e8panse** [comment]
There gists:

[https://gist.github.com/e8panse/8d64d6c8a8d4b0390c6f5750529d752c](url)

[https://gist.github.com/e8panse/d94e75fa838bda979fadd2491957a6e3](url)

**Comment by tilakrayal** [comment]
@e8panse,
Though the code was provided, it was not indentation in the way to execute. I tried to make it run with the changes but I was not able to execute. Kindly find the [gist](https://colab.research.google.com/gist/tilakrayal/c55adcbda4441999d3d882760b8881d2/untitled910.ipynb) and provide the code with indentation which helps us to debug the issue. Thank you!
